llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -1876,6 +1876,47 @@ static Value *simplifyInstructionWithPHI(Instruction &I, PHINode *PN,
   return nullptr;
 }
 
+// In some cases it is beneficial to fold a select into a binary operator.
+// For example:
+//   %1 = or %in, 4
+//   %2 = select %cond, %1, %in
+//   %3 = or %2, 1
+// =>
+//   %1 = select i1 %cond, 5, 1
+//   %2 = or %1, %in
+Instruction *InstCombinerImpl::foldSelectIntoBinOp(BinaryOperator &Op) {
+  SelectInst *SI = dyn_cast<SelectInst>(Op.getOperand(0));
+  ConstantInt *Const = dyn_cast<ConstantInt>(Op.getOperand(1));
```

Comment:
Can you add some vector tests? ConstantInt cannot represent vector constants. I'd suggest using the m_ImmConstant matcher.

Specifically, please add a case where the condition of select is a scalar while the arms are vectors.